docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,317 @@
+    *  @param rx_buffer_size The size of the receive buffer
+    *  @param tx_buffer The transmit buffer
+    *  @param tx_buffer_size The size of the transmit buffer
+    */
+    BufferedSerial(PinName tx,
+                   PinName rx,
+                   int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE,
+                   void *rx_buffer,
+                   size_t rx_buffer_size,
+                   void *tx_buffer,
+                   size_t tx_buffer_size);
+```
+
+ The rest of the class API is that of `UARTSerial` and is described [here](https://os.mbed.com/docs/mbed-os/v5.14/mbed-os-api-doxy/classmbed_1_1_u_a_r_t_serial.html).
+
+
+### Detailed design : UnbufferedSerial
+
+`UnbufferedSerial` is a new class that adds a `FileHandle` interface to `RawSerial`. `RawSerial` is the preferred implementation for unbuffered I/O but it lacks a `FileHandle` interface and so it cannot be used by the retarget code. Instead the retarget code uses a non-public `DirectSerial` class which adds to the complexity of classes and should be removed. `UnbufferedSerial` will not provide any of the printf methods that `RawSerial` has; applications should use _fdopen_ to get a _FILE*_ to use the printf function, see [Usage scenarios and examples](#usage-scenarios-and-examples).
+
+The below diagram shows the detailed inheritance hierarchy for `UnbufferedSerial`.
+
+![UnbufferedSerial class](./diagrams/unbuffered-serial.png)
+
+**API description**
+
+This is the proposed API for the `UnbufferedSerial` class.
+
+```C
+class UnbufferedSerial: public SerialBase, public FileHandle, private NonCopyable<UnbufferedSerial> {
+
+public:
+    /** Create a UnbufferedSerial port, connected to the specified transmit and receive pins, with the specified baud.
+     *
+     *  @param tx Transmit pin
+     *  @param rx Receive pin
+     *  @param baud The baud rate of the serial port (optional, defaults to MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE)
+     *
+     *  @note
+     *    Either tx or rx may be specified as NC if unused
+     */
+    UnbufferedSerial(PinName tx, PinName rx, int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE);
+
+    virtual ssize_t write(const void *buffer, size_t size);
+    virtual ssize_t read(void *buffer, size_t size);
+    virtual off_t seek(off_t offset, int whence = SEEK_SET);
+    virtual off_t size();
+    virtual int isatty();
+    virtual int close();
+    virtual short poll(short events) const;
+
+#if !(DOXYGEN_ONLY)
+protected:
+
+    /* Acquire exclusive access to this serial port
+     */
+    virtual void lock(void);
+
+    /* Release exclusive access to this serial port
+     */
+    virtual void unlock(void);
+#endif
+};
+```
+
+# Usage scenarios and examples
+
+### Scenario 1 `Printing to stdout`
+
+There is no change in the current APi in regard to printing to stdout.  Applications should call `printf` directly as shown below. The console baud rate can be configured via `platform.stdio-baud-rate` in `mbed_app.json`.
```

Comment:
I don't really like the `should`, I would say `can` and also add a small disclaimer, something like `(it can negatively impact the image size)`

Reply:
Negatively impact compared to what? I believe this section started as a comparator to `Serial::printf`, and I think that's its focus. The example doesn't even need to be `printf`, it could have been `puts`.

If we want to talk about `printf` versus alternatives, that's at a bit of a tangent to the serial work, but I guess the possibilities are:

* Using `printf`
* Using `printf` with `minimal-printf` on, reducing formatter
* Only using `puts` (or all your `printf`s are simple enough for compiler to turn them into `puts`), avoiding formatter
* Using just `write`, avoiding C library altogether
* Turning on `minimal-console` (in conjunction with any of the 4 previous options)

Reply:
I changed `should` into `can` and put the example using `write` first. Discussion related to printf comes after so I am not emphasing the use of printf.